pennylane/math/multi_dispatch.py

```diff
@@ -0,0 +1,340 @@
+        # to reproduce numpy's behaviour
+        values = [np.flatten(np.array(t)) for t in values]
+        axis = 0
+
+    return np.concatenate(values, axis=axis, like=interface)
+
+
+def diag(values, k=0):
+    """Construct a diagonal tensor from a list of scalars.
+
+    Args:
+        values (tensor_like or Sequence[scalar]): sequence of numeric values that
+            make up the diagonal
+        k (int): The diagonal in question. ``k=0`` corresponds to the main diagonal.
+            Use ``k>0`` for diagonals above the main diagonal, and ``k<0`` for
+            diagonals below the main diagonal.
+
+    Returns:
+        tensor_like: the 2D diagonal tensor
+
+    **Example**
+
+    >>> x = [1., 2., tf.Variable(3.)]
+    >>> diag(x)
+    <tf.Tensor: shape=(3, 3), dtype=float32, numpy=
+    array([[1., 0., 0.],
+           [0., 2., 0.],
+           [0., 0., 3.]], dtype=float32)>
+    >>> y = tf.Variable([0.65, 0.2, 0.1])
+    >>> diag(y, k=-1)
+    <tf.Tensor: shape=(4, 4), dtype=float32, numpy=
+    array([[0.  , 0.  , 0.  , 0.  ],
+           [0.65, 0.  , 0.  , 0.  ],
+           [0.  , 0.2 , 0.  , 0.  ],
+           [0.  , 0.  , 0.1 , 0.  ]], dtype=float32)>
+    >>> z = torch.tensor([0.1, 0.2])
+    >>> qml.math.diag(z, k=1)
+    tensor([[0.0000, 0.1000, 0.0000],
+            [0.0000, 0.0000, 0.2000],
+            [0.0000, 0.0000, 0.0000]])
+    """
+    interface = _multi_dispatch(values)
+
+    if isinstance(values, (list, tuple)):
+        values = np.stack(np.coerce(values, like=interface), like=interface)
+
+    return np.diag(values, k=k, like=interface)
+
+
+def dot(tensor1, tensor2):
+    """Returns the matrix or dot product of two tensors.
+
+    * If both tensors are 0-dimensional, elementwise multiplication
+      is performed and a 0-dimensional scalar returned.
+
+    * If both tensors are 1-dimensional, the dot product is returned.
+
+    * If the first array is 2-dimensional and the second array 1-dimensional,
+      the matrix-vector product is returned.
+
+    * If both tensors are 2-dimensional, the matrix product is returned.
+
+    * Finally, if the the first array is N-dimensional and the second array
+      M-dimensional, a sum product over the last dimension of the first array,
+      and the second-to-last dimension of the second array is returned.
+
+    Args:
+        tensor1 (tensor_like): input tensor
+        tensor2 (tensor_like): input tensor
+    """
+    interface = _multi_dispatch([tensor1, tensor2])
+    x, y = np.coerce([tensor1, tensor2], like=interface)
+
+    if interface == "torch":
+        if x.ndim == 0 and y.ndim == 0:
+            return x * y
+
+        if x.ndim <= 2 and y.ndim <= 2:
+            return x @ y
+
+        return np.tensordot(x, y, dims=[[-1], [-2]], like=interface)
+
+    if interface == "tensorflow":
+        if x.ndim == 0 and y.ndim == 0:
+            return x * y
+
+        if y.ndim == 1:
+            return np.tensordot(x, y, axes=[[-1], [0]], like=interface)
+

```

Comment:
Is this a very specific TF issue?

Reply:
Yep. Every single framework treats `tensordot` differently. It is _very_ annoying.

Reply:
oof

Reply:
It also highlights that `autoray` is more of a single dispatch library than a full compatibility layer